icu4c/source/i18n/unitconverter.cpp

```diff
@@ -5,14 +5,494 @@
+    } else if (baseStr == "PI") {
+        factor.constants[CONSTANT_PI] += power * signal;
+    } else {
+        if (signal == Signal::NEGATIVE) {
+            factor.factorDen *= std::pow(strToDouble(baseStr), power);
+        } else {
+            factor.factorNum *= std::pow(strToDouble(baseStr), power);
+        }
+    }
+}
+
+/*
+  Adds single factor for a `Factor` object. Single factor means "23^2", "23.3333", "ft2m^3" ...etc.
+  However, complext factor are not included, such as "ft2m^3*200/3"
+*/
+void addFactorElement(Factor &factor, StringPiece elementStr, Signal signal, UErrorCode &status) {
+    StringPiece baseStr;
+    StringPiece powerStr;
+    int32_t power =
+        1; // In case the power is not written, then, the power is equal 1 ==> `ft2m^1` == `ft2m`
+
+    // Search for the power part
+    int32_t powerInd = -1;
+    CharString charStr(elementStr, status);
+    for (int32_t i = 0, n = charStr.length(); i < n; ++i) {
+        if (charStr[i] == '^') {
+            powerInd = i;
+            break;
+        }
+    }
+
+    if (powerInd > -1) {
+        // There is power
+        baseStr = elementStr.substr(0, powerInd);
+        powerStr = elementStr.substr(powerInd + 1);
+
+        power = static_cast<int32_t>(strToDouble(powerStr));
+    } else {
+        baseStr = elementStr;
+    }
+
+    addSingleFactorConstant(factor, baseStr, power, signal);
+}
+
+/*
+ * Extracts `Factor` from a complete string factor. e.g. "ft2m^3*1007/cup2m3*3"
+ */
+void extractFactor(Factor &factor, StringPiece stringFactor, UErrorCode &status) {
+    // Set factor to `1`
+    factor.factorNum = 1;
+    factor.factorDen = 1;
+
+    Signal signal = Signal::POSITIVE;
+    auto factorData = stringFactor.data();
+    for (int32_t i = 0, start = 0, n = stringFactor.length(); i < n; i++) {
+        if (factorData[i] == '*' || factorData[i] == '/') {
+            StringPiece factorElement = stringFactor.substr(start, i - start);
+            addFactorElement(factor, factorElement, signal, status);
+
+            start = i + 1; // Set `start` to point to the start of the new element.
+        } else if (i == n - 1) {
+            // Last element
+            addFactorElement(factor, stringFactor.substr(start, i + 1), signal, status);
+        }
+
+        if (factorData[i] == '/')
+            signal = Signal::NEGATIVE; // Change the signal because we reached the Denominator.
+    }
+}
+
+// Load factor for a single source
+void loadSingleFactor(Factor &factor, StringPiece source, UErrorCode &status) {
+    auto conversionUnit = extractConvertUnit(source, status);
+    if (U_FAILURE(status)) return;
+
+    extractFactor(factor, conversionUnit.factor, status);
+    factor.offset = strHasDivideSignToDouble(conversionUnit.offset, status);
+    factor.reciprocal = factor.reciprocal;
+}
+
+// Load Factor for compound source
+void loadCompoundFactor(Factor &factor, StringPiece source, UErrorCode &status) {
```

Comment:
Is there a typical order of parameters in ICU? Stylistically I'm used to inputs before outputs...

When reading the call sites with no cues, it's surprising to me to see it in this order. But I'm not familiar with ICU.

A bit off-topic, for comparison:
Google's style has inputs before outputs. That style also discourages references for output though, which is definitely not the case for ICU style. (What's nice about pointers as output is that it's easier to see what's an output at the callsite. But then nowadays IDEs help by showing you the function fingerprint e.g. when you hover on a symbol...)
https://google.github.io/styleguide/cppguide.html#Output_Parameters
https://google.github.io/styleguide/cppguide.html#Reference_Arguments

Reply:
Yeah, `factor` and `source` are probably out of order.  Not a huge deal since this is a private internal function, but maybe worth fixing.

Reply:
it returns a `Factor` now

Reply:
and I have changed the others.